flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ProgramDeployer.java

```diff
@@ -18,163 +18,83 @@
 		if (LOG.isDebugEnabled()) {
 			LOG.debug("Submitting job {} with the following environment: \n{}",
-					jobGraph.getJobID(), context.getMergedEnvironment());
+					pipeline, context.getMergedEnvironment());
 		}
-		deployJob(context, jobGraph, result);
-	}
 
-	public JobExecutionResult fetchExecutionResult() {
-		return executionResultBucket.poll();
-	}
+		// create a copy so that we can change settings without affecting the original config
+		Configuration configuration = new Configuration(context.getFlinkConfig());
+		if (configuration.get(DeploymentOptions.TARGET) == null) {
+			throw new RuntimeException("No execution.target specified in your configuration file.");
+		}
 
-	/**
-	 * Deploys a job. Depending on the deployment creates a new job cluster. It saves the cluster id in
-	 * the result and blocks until job completion.
-	 */
-	private <T> void deployJob(ExecutionContext<T> context, JobGraph jobGraph, Result<T> result) {
-		// create or retrieve cluster and deploy job
-		try (final ClusterDescriptor<T> clusterDescriptor = context.createClusterDescriptor()) {
-			try {
-				// new cluster
-				if (context.getClusterId() == null) {
-					deployJobOnNewCluster(clusterDescriptor, jobGraph, result, context.getClassLoader());
-				}
-				// reuse existing cluster
-				else {
-					deployJobOnExistingCluster(context.getClusterId(), clusterDescriptor, jobGraph, result);
-				}
-			} catch (Exception e) {
-				throw new SqlExecutionException("Could not retrieve or create a cluster.", e);
-			}
-		} catch (SqlExecutionException e) {
-			throw e;
-		} catch (Exception e) {
-			throw new SqlExecutionException("Could not locate a cluster.", e);
+		if (awaitJobResult) {
+			configuration.set(DeploymentOptions.ATTACHED, true);
 		}
-	}
 
-	private <T> void deployJobOnNewCluster(
-			ClusterDescriptor<T> clusterDescriptor,
-			JobGraph jobGraph,
-			Result<T> result,
-			ClassLoader classLoader) throws Exception {
-		ClusterClient<T> clusterClient = null;
+		ExecutorServiceLoader executorServiceLoader = DefaultExecutorServiceLoader.INSTANCE;
+		final ExecutorFactory executorFactory;
 		try {
-			// deploy job cluster with job attached
-			clusterClient = clusterDescriptor.deployJobCluster(context.getClusterSpec(), jobGraph, false);
-			// save information about the new cluster
-			result.setClusterInformation(clusterClient.getClusterId(), clusterClient.getWebInterfaceURL());
-			// get result
-			if (awaitJobResult) {
-				// we need to hard cast for now
-				final JobExecutionResult jobResult = clusterClient
-						.requestJobResult(jobGraph.getJobID())
-						.get()
-						.toJobExecutionResult(context.getClassLoader()); // throws exception if job fails
-				executionResultBucket.add(jobResult);
-			}
-		} finally {
-			try {
-				if (clusterClient != null) {
-					clusterClient.close();
-				}
-			} catch (Exception e) {
-				// ignore
-			}
+			executorFactory = executorServiceLoader.getExecutorFactory(configuration);
```

Comment:
Before we were also closing the `clusterClient` here. Now we do not because we return the `JobClient`, so it becomes the responsibility of the call-site to close it but, in general, we do not do it so we may have resource-leakage.

Reply:
In this ongoing commit https://github.com/apache/flink/commit/daf85a75b9c24918058b8bfe09416b2828bd02a5 I introduce an approach that we can add a hook closing cluster client on job client closed. Since we return job client it is up to the caller when close it. We just need to ensure that the underneath cluster client has also been closed because it is internal implementation details.